src/components/tab-panel/tab-panel.tsx

```diff
@@ -0,0 +1,93 @@
+import {
+    Component,
+    Element,
+    h,
+    Prop,
+    EventEmitter,
+    Event,
+    Host,
+} from '@stencil/core';
+import { Tab } from '../tab-bar/tab.types';
+
+/**
+ * @slot - Content to put inside the `tab-panel`. Each slotted element must have the `id` attribute equal to the id of the tab it belongs to.
+ */
+@Component({
+    tag: 'limel-tab-panel',
+    shadow: true,
+})
+export class TabPanel {
+    @Prop({ mutable: true })
```

Comment:
Does this really have to be mutable?

Reply:
No, but it simplifies the code **alot**

My first implementation was without it being mutable and it forces us to have two arrays of the tabs and keep them in sync. Since we have to listen for the `changeTab` event to be able to display the active content (and also if the event comes from within the inner components), we must trigger a re-render. If the consumer does not add a `changeTab` listener and updates the tabs accordingy, it will break this component.

I think mutable might sound worse than it actually is. It will not mutate the consumer's reference to the tabs, only our local copy inside this component. When/if the consumer updates their reference, this one will be updated accordingly as well

Reply:
> It will not mutate the consumer's reference to the tabs, only our local copy inside this component.

Ah, that's nice, I didn't know that 👍

I was partially asking for "should we update it with the event-handler-loop we usually do" part that you answered, but I was also partially asking because I wasn't sure if a "non-mutable" Prop decorator actually stopped internal changes to the *content* of the array, or only stopped internal replacing of the whole object itself. If all we're doing is updating properties on the objects contained in the array, I'd be pretty surprised if that required the `mutable` flag, but maybe it does?

Reply:
I don't think it will stop us changing the contents of the array, but that would also not trigger a re-render. Also, if we were to mutate the actual items in the array, that would probably also update the references that the consumer is holding onto :dizzy_face:

Reply:
Have not tested that though, but it seems like a really bad practice :smile:

Reply:
Ah! Yeah, of course! I didn't think of that… 😳

👌